src/resolver.rs

```diff
@@ -79,56 +87,144 @@ impl Resolver {
+                        meta,
+                        &mut deactivated,
+                    );
+                    // 2. set latest version
+                    if current_version_id < block_number {
+                        current_version_id = block_number;
+                    }
+                } else {
+                    // just update the next version before quitting.
+                    last_updated_did_version_id = Some(block_number);
+                    break;
                 }
-                DIDRegistryEvents::DidownerChangedFilter(owner_changed) => {
-                    base_document.owner_event(owner_changed)
+            } else {
+                // 1. delegate events
+                Resolver::dispatch_event(
+                    self,
+                    &mut base_document,
+                    public_key,
+                    event,
+                    meta,
+                    &mut deactivated,
+                );
+                // 2. set latest version
+                if current_version_id < block_number {
+                    current_version_id = block_number;
                 }
             };
-
-            // if a did was set with the wrong format, for instance set-attribute was called with
-            // raw bytes instead of hex-encoded bytes, we don't want to cancel resolution of the
-            // rest of the DID
-            //
-            // TODO: Send this info as an extra json field to the caller apart from the DID Document
-            if let Err(e) = res {
-                log::error!(
-                        "Error while resolving for {public_key} at event block={block_number}, log index={log_index}, incorrect format?: {e}",
-                    );
-            };
         }
-        Ok(base_document.build())
+
+        // get the timestamp for the current_verison_id
+        let current_version_timestamp = match self.signer.get_block(current_version_id).await? {
+            Some(block) => {
+                if current_version_id != U64::zero() {
+                    Some::<String>(
+                        block
+                            .time()
+                            .unwrap_or_default()
+                            .format("%Y-%m-%dT%H:%M:%SZ")
+                            .to_string(),
+                    )
+                } else {
+                    None::<String>
+                }
+            }
+            None => None::<String>,
+        };
+
+        let resolution_result = DidResolutionResult {
+            document: base_document.build(),
+            metadata: Some(DidDocumentMetadata {
+                deactivated,
+                version_id: current_version_id.as_u64(),
+                updated: current_version_timestamp,
+                next_version_id: last_updated_did_version_id.map(|ver| ver.as_u64()),
+                next_update: match last_updated_did_version_id {
+                    Some(ver) => self.signer.get_block(ver).await?.map(|block| {
```

Comment:
since we use the same code for getting the time from a block, we can define a closure to do it for us which will clean up both sections of code a bit:

```rust
let block_time = |block: Block<H256>| {
    block
        .time()
        .unwrap_or_default()
        .format("%Y-%m-%dT%H:%M:%SZ")
        .to_string()
};
```

it then becomes
```rust
next_update: match last_updated_did_version_id {
            Some(ver) => self.signer.get_block(ver).await?.map(block_time),
            None => None::<String>,
        },
```

and for the previous bit,

```rust
let current_version_timestamp = self
    .signer
    .get_block(current_version_id)
    .await?
    .map(block_time);
```